### ISSUE TITLE
Allow @Default to overrule potentially existing parameterless default constructor.

### ISSUE THREAD

**Issue by ralfhecktor** [mapstruct/mapstruct#2150]
Since 1.4.0.Beta2, MapStruct allows direct mapping to parameterized constructors as described [here](https://mapstruct.org/documentation/dev/reference/html/#mapping-with-constructors). The reference guide states:

> * If a parameterless constructor exists then it would be used to construct the object, and the other constructors will be ignored
> * If there are multiple constructors then the one annotated with annotation named @Default (from any package) will be used

I understand this as follows: If there is a parameterless default constructor, this one wins and @Default is only interpreted if there is no paramterless default constructor. However, this is problematic in a setup, where a default constructor is only made visible to the objects children, so being protected. Lets take the following example:

``` java
public class Vehicle {
    
    private String color;
    private int amountOfWheels;
    
    protected Vehicle() {
        // e.g. needed for serialization.
    }

    @Default
    public Vehicle(String color, int amountOfWheels) {
        this.color = color;
        this.amountOfWheels = amountOfWheels;
    }
...
}
```
and
``` java
public class Car extends Vehicle {

    private String brand;

    private Car() {
        // e.g. needed for serialization.
    }

    public Car(String brand, Vehicle vehicle) {
        super(vehicle.getColor(), vehicle.getAmountOfWheels());
        this.brand = brand;
    }
...
}
```
with a DTO as follows
``` java
public class VehicleDTO {

    private String color;
    private int amountOfWheels;

...
}
```
and the according mapper
``` java
@Mapper
public interface VehicleMapper {

    VehicleMapper INSTANCE = Mappers.getMapper(VehicleMapper.class);

    Vehicle map(VehicleDTO dto);
}
```

In this case I want the default constructor to be only visible to the child as it is needed e.g. for serialization and want MapStruct to ignore it and use the declared @Default constructor. However, MapStruct is picking up the protected default constructor and throws the following error:

> Compilation failure
> [ERROR] MapStructsExample/target/generated-sources/annotations/de/ralfhecktor/mapper/VehicleMapperImpl.java:[20,27] constructor Vehicle in class de.ralfhecktor.Vehicle cannot be applied to given types;
> [ERROR]   required: java.lang.String,int
> [ERROR]   found: no arguments
> [ERROR]   reason: actual and formal argument lists differ in length

Having a look at the generated implementation shows:

``` java
public class VehicleMapperImpl implements VehicleMapper {

    @Override
    public Vehicle map(VehicleDTO dto) {
        if ( dto == null ) {
            return null;
        }

        Vehicle vehicle = new Vehicle();

        return vehicle;
    }
}
```
instead of 

``` java
public class VehicleMapperImpl implements VehicleMapper {

    @Override
    public Vehicle map(VehicleDTO dto) {
        if ( dto == null ) {
            return null;
        }

        String color = null;
        int amountOfWheels = 0;

        color = dto.getColor();
        amountOfWheels = dto.getAmountOfWheels();

        Vehicle vehicle = new Vehicle( color, amountOfWheels );

        return vehicle;
    }
}
```

So what would be suggestion on solving this? 

For MapStruct I would see three possibilities:

1. Only consider `public` constructor to be preferred over the @Default.
1. Make @Default a definite override, so if somebody declares a constructor as default, it should also be picked up.
1. Give @Default a property e.g. `force` which can be set to `true` to force the usage.

Personally, I would find option 2 the most intuitive. 

**You can find a full working project [here](https://github.com/ralfhecktor/MapStructExample)**

I am looking forward to your thoughts and suggestions on this topic. Thanks.

**Comment by filiphr** [comment]
Wow this is a great detailed issue. Thanks a lot for the detailed explanation, examples and the reproducible project.

Thinking a bit more on this I think taking 2 of your possibilities in the default setup is the way to go:

* If there is a single `public` constructor we should use that.
* If a constructor is annotated with `@Default` then use that one no matter what

The third possibility is no longer valid, since `@Default` is a definitive override.

For your use case if you have a single `public` constructor you won't need `@Default`.

Basically the new rules should be:

* If there is a constructor (public or protected) annotated with an annotation named `@Default` (from any package) it will be used
* If there is a single public constructor then it would be used to construct the object
* If a parameterless constructor exists then it would be used to construct the object, and the other constructors will be ignored

What do you think @sjaakd?
